counter/src/main/java/org/infinispan/counter/impl/weak/WeakCounterImpl.java

```diff
@@ -128,14 +128,19 @@ public static CompletionStage<Void> removeWeakCounter(Cache<WeakCounterKey, Coun
     * <p>
     * Only one key will have the initial value and the remaining is zero.
     */
-   public void init() {
+   public CompletionStage<InternalCounterAdmin> init() {
       registerListener();
-      for (int i = 0; i < entries.length; ++i) {
-         final int index = i;
-         awaitCounterOperation(readOnlyMap.eval(entries[index].key, ReadFunction.getInstance())
-               .thenAccept(value -> initEntry(index, value)));
+      AggregateCompletionStage<InternalCounterAdmin> stage = CompletionStages.aggregateCompletionStage(this);
+      synchronized (entries) {
```

Comment:
The `entries` variable is not protected in all of the places. Shouldn't all of the places do so? For example `KeySelector#updatePreferredKeys` references it without synchronization in a few places.